File: lex.py
p_reservadas = ["class", "else", "false", "fi", "if", "in", "inherits", "isvoid", "let", "loop", "pool", 
                "then", "while", "case", "esac", "new", "of", "not", "true"]

lst_read = None

def lexico(arq):
    global lst_read
    palavra = "" 

    if lst_read == ";":
        palavra += lst_read
        return palavra
    
    while True:
        c = arq.read(1)
        lst_read = c
        if not c:
            return palavra if palavra else None
            
        if c.isalpha():
            palavra += c
        
        elif c.isspace():
            if palavra:
                lst_read = None
                return palavra
            continue

        elif c == ";":
            if palavra:
                return palavra
            return ";"
        
        elif c == "-":
            lst_read = c
            c = arq.read(1)
            if c == "-":
                while True:
                    c = arq.read(1)
                    if c == "-":
                        c = arq.read(1)
                        if c == "-":
                            break
                    else:
                        continue
            elif c.isalpha():
                palavra += lst_read + c



def main():
    with open("teste.txt", "r") as arquivo:
        # Primeira chamada
        p1 = lexico(arquivo)
        print(f"Token 1: {p1}")
        
        # Segunda chamada (continua de onde parou!)
        p2 = lexico(arquivo)
        print(f"Token 2: {p2}")
main()

File: lex.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

typedef struct token
{
    char simbol [20];
    char tipo; // se vai ser palavra reservada, constante, variavel, etc...

}token; // talvez fazer isos depois

void lexico2(FILE* arq) {
    char palavra[100];
    int i = 0;
    char c;

    while ((c = fgetc(arq)) != EOF){
        if (isalpha(c)) {
            palavra[i] = c;
            i++;
        }
        /*else if (i > 0) { 
            palavra[i] = '\0'; // Finaliza a string com o caractere nulo
            printf("Palavra encontrada: %s\n", palavra);
        }*/
        if(isspace(c)){
            palavra[i] = '\0';
            printf("Palavra encontrada: %s\n", palavra);
            break;
        }
    }
}

int main(){
    FILE* arquivoCOOL = fopen("teste.txt", "r");
    lexico(arquivoCOOL);

    return 0;
}